Port CesiumUpload Tool to Node

ggetz opened the pull request:
Added gulp task for uploading cesium. Port of the CesiumUploader tool.

The gulp task is named `deploy` and requies two arguments:
```
-b <BucketName>
-d <UploadDirectoryName>
```

This is encapsulated in the npm script `uploadCesium`, which runs `gulp deploy -b cesiumjs -d cesium`

I use the `aws-sdk` node module, and configure `maxRetries` to `10` and `retryDelayOptions.base` to `500` because I was getting `ECONNRESET` reset errors when uploading a large amount of files. This uses the credentials specified in the `~/.aws` directory.

I also used the `graceful-fs` module because there were too many files open at once, and this module handles the number open at once. I used `zlib` to for compression and `crypto` for generating the hash for each file. I also used `mime` to determine the mime type of each file, and `compressible` to determine if the files are compressible. Let me know if there is a better option for any of these.

I'm still having a little trouble with the delete step. If there is a large amount of files (about 1000+) that need to be deleted, I get an error `MalformedXML` when making the request to s3. I think the list of files in the single delete request is to large. Should I make each request individually?

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,217 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+            var contentEncoding = compress ? 'gzip' : undefined;
+            var etag;
+
+            return graceful.readFile(filename, function(err, contents) {
```
There's no need to use `graceful-fs` here, it actually masks the real problem and will hurt overall performance.  Instead, you should limit the amount of concurrency by using `async.forEachLimit` and passing it a reasonable number of files (say 2500), so that it doesn't try and open too many files as once.

ggetz commented:
I removed `graceful-fs` and set the concurrency limit to 2000, and it is indeed faster.

mramato commented:
Good to hear, I'll take a closer look at this soon.  Is this branch actually deploying to s3 now?

ggetz commented:
Not in Travis, no. I don't have credentials that have all the correct access to the "cesium-dev" bucket.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +302,218 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+            dot: true // include hidden files
+        });
+
+        return async.forEachLimit (files, concurrencyLimit, function(filename, callback) {
```
The mix of callbacks and promises here gets pretty confusing.  I think we can easily move to an all promises based approach.  Instead of `async.forEachLimit` we can use [Promise.map](http://bluebirdjs.com/docs/api/promise.map.html) with a concurrency of 2000.  Also, I believe globby returns a promis as well.  Finally, you can turn `fs.readFile`, `zlib.gzip` (and any other callback functions) into a  Promise-based function by calling `var fsReadFile = Promise.promisify(fs.readFile)`

This should make the code much cleaner all around and completely avoid the need for `new Promise` or callbacks anwhere.

ggetz replied:
I haven't had to deal much with promises before, so I'm sorry for the confusing approach!

mramato replied:
No need to apologize, that's why we do code reviews.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +302,218 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+    var uploaded = 0;
+
+    // get all files currently in bucket
+    function listAll(callback, marker) {
```
Rather than make functions inside the task, declare them ourtide of the tasks but with more descriptive names.  We might eventually break this master gulpfile.js up into small modules for easier maintenance, but that can be done in a future PR.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +302,218 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+
+    var uploadDir = argv.d;
+    var bucketName = argv.b;
+
```
We should prompt for confirmation with the provided details before doing anything.  The readline module (which is included with node) can be used for this:

```javascript
    var iface = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    iface.question("Question", function (answer) {

    });
```

mramato commented on gulpfile.js:
```diff
@@ -297,6 +302,218 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+        return async.forEachLimit (files, concurrencyLimit, function(filename, callback) {
+            var blobName = uploadDir + '/' + filename;
+            var contentType = mime.lookup(filename);
+            var compress = compressible(contentType);
```
I don't think either the `mime` or `compressible` modules have the same behavior as the simple class that we had in the old tool.  For example, does this correctly detect terrain, geojson and czml?  Assuming it's just using the extension, I highly doubt it.  While I'm all for using third-party libraries, since we only have a few types to worry about it might be easier to just port the old behavior directly.

mramato replied:
Of course if I'm wrong and it does correctly detect everything, then maybe we can just stick with this.  We just have to be mindful of depending on too many third-party modules, especially since many are poorly maintained.

ggetz replied:
I know with `mime` you can [define new types](https://github.com/broofa/node-mime#mimedefine), but I'm fine with switching to just porting the old behavior directly if we want to avoid third-party libraries.

mramato commented:
Thanks @ggetz.  Since I suggested some significant changes, that's all I have for now. I think this looks pretty good and will work really well.  Let me knwo when it's ready for another look.

ggetz commented:
@mramato Ready for another look. 

I got rid of the callbacks and made everything a promise. I also replaced the `aws-sdk` module with `aws-sdk-promise`, which is the same as the previous module, but replaces the callbacks with promises. 

I added the prompt for confirmation with `readline`. However the process hangs after the gulp task has finished executing, so I think the `readline` process is still running. Did I use this right?

Also what should we do about the prompt for Travis? Should it use a separate task without a prompt, or should we add a force flag?

Lastly, I added the checks for special mime types, but I left in `mime` and `compressible` for the rest of the file types not defined in the old tool.

mramato commented:
@ggetz looking at `aws-sdk-promise` it's not an official aws library and has comments like `This is a terrible hack but it should not break anything in horrible`.  The whole module is also under 10 lines of code.  I would remove it and write a helper function to replace it.

mramato commented:
> Also what should we do about the prompt for Travis? Should it use a separate task without a prompt, or should we add a force flag?

Just add a --no-prompt flag and have travis.yml use it.

> Lastly, I added the checks for special mime types, but I left in mime and compressible for the rest of the file types not defined in the old tool.

This sounds good for now, `mime` seems to be a pretty standard module.

ggetz commented:
> looking at aws-sdk-promise it's not an official aws library and has comments like This is a terrible hack but it should not break anything in horrible. The whole module is also under 10 lines of code. I would remove it and write a helper function to replace it.

OK

ggetz commented:
I promisified all the `s3` functions and added a `confirm` flag which skips the confirmation prompt because yargs did not properly capture flags with hyphens in the name.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,216 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+
+    return listAll(s3, bucketName, uploadDirectory, existingBlobs)
+    .then(function() {
+        return globby([
```
We are uploading way more than we need to.  It would probably be better to use a specific inclusion list rather than including everything and then excluding.  Here's what the official 1.19 release uploaded looks like:

![image](https://cloud.githubusercontent.com/assets/1495196/13964225/187df6b6-f03f-11e5-9cb6-d4e2a185de1d.png)

We should just start with that.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,216 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+    } else {
+        // prompt for confirmation
+        iface.question('Files from your computer will be published to the ' + bucketName + ' bucket. Continue? [y/n] ', function(answer) {
+            if (answer === 'y') {
```
Add `iface.close();` before this line, that's what preventing the script from exiting.

ggetz replied:
I added `iface.close();` and it's still not ending the process.

mramato replied:
It's possible you have multiple issues then and this was just one of them.  I'll take a look once everything else is ready.

ggetz replied:
OK, thanks!

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,216 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+        deployCesium(bucketName, uploadDirectory).then(function() {
+            done();
+        });
+    } else {
```
Just return in the if above and you can get rid of the else clause.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,216 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+    var readFile = Promise.promisify(fs.readFile);
+    var gzip = Promise.promisify(zlib.gzip);
+    var concurrencyLimit = 2000;
+    var cacheControl = "max-age=3600";
```
For actual deployment, this is what we want, but for CI it doesn't make sense because changes won't be automatically reflected if the user doesn't clear their cache.  The solution is to make this an optional command line argument and when we do travis deploys we will use `no-cache` as the setting.

mramato commented:
Thanks, @ggetz this is coming a long great.  Just those last few comments.

ggetz commented:
@mramato I made the updates. CacheControl can be specified with an option `-c` flag.

mramato commented on package.json:
```diff
@@ -90,6 +93,7 @@
     "generateStubs": "gulp generateStubs",
-    "sortRequires": "gulp sortRequires"
+    "sortRequires": "gulp sortRequires",
+    "uploadCesium": "gulp deploy -b cesiumjs -d cesium"
```
As we discussed offline, rename the npm and gulp tasks to  `deploy-s3`, also don't hardcode the options here.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,224 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+                            contents = result;
+                        });
+                    } else {
+                        contents = content;
```
This is pretty confusing, we should be able to just `return compress ? gzip(content) : content;` and get rid of the `contents` object up top.  Then instead of returning true/false below, either return content or undefined.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,224 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+                    etag = crypto.createHash('md5').update(contents).digest('base64');
+
+                    var index = existingBlobs.indexOf(blobName);
+                    if (index > -1) {
```
Always favor early return, that is, check for `index <= -1` and return true immediately instead of having a big chunk of code in the if block.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,224 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+                            Key : blobName
+                        })
+                        .then(function(data) {
+                            if (data.ETag === ('"' + hash + '"') &&
```
Same comment.

mramato replied:
I won't mention it any more, but throughout change everything to early return.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,224 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+                        return s3.putObjectAsync(params).then(function() {
+                            uploaded++;
+                        },
+                        function(error) {
```
Always use `catch` instead of a second parameter to `then`.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,224 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+}
+
+// delete unneeded files in bucket asynchronously
+function cleanFiles(s3, bucketName, files) {
```
You can get rid of this function and just call this up top:

```javascript
                return s3.deleteObjectsAsync({
                           Bucket : bucketName,
                           Delete : {
                               Objects : existingBlobs.map(function(file) {
                                   return {Key : file};
                               })
                           }
                       });
```

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,224 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+
+                    var index = existingBlobs.indexOf(blobName);
+                    if (index > -1) {
+                        // remove files ad we find them on disk
```
I assume `ad` -> `as`?

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,224 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+            return listAll(s3, bucketName, directory, files, files[files.length - 1]);
+        }
+    },
+    function(error) {
```
Use catch instead.

mramato commented:
This is very close, just those last few comments.

ggetz commented:
Updated the task and script names and made the other changes you suggested.

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,216 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+                    return undefined;
+                })
+                .catch(function() {
+                    return content;
```
I'm having a hard time figuring out how this catch block ever gets triggered.  The only case I can think of is if a file is in `existingBlobs` but is deleted by something else before we get to the `headObjectAsync` call.  Is there another case I'm not thinking of?  Part of me things we should just take it out since it might mask another error.  Why is it needed?

ggetz replied:
I figured I'd have it there just in case anything went wrong with that request, the process could still continue. For example, before I configured the `maxRetries` and `retryDelay` time, I was getting a lot of errors in this step even though the file exists when making the `headObjectAsync` call.

mramato replied:
I would just change this to console.log(e) instead, otherwise we are masking errors.

Actually, now that I think about it.  Create an array called `errors` and anywhere you are console logging an error, push it onto the array instead.  Then at the end, fail the task and print out the errors if the array has anything in it.

ggetz replied:
OK

mramato commented on gulpfile.js:
```diff
@@ -297,6 +303,216 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+                return s3.putObjectAsync(params).then(function() {
+                    uploaded++;
+                })
+                .catch(function(error) {
```
This catch should go one level down because if `readFile` or any of the other blocks cause an error, it will be silently ignored.  By moving it to the parent promise chain, it will catch any and all errors.

mramato replied:
Actually, never mind.  Leave this here.  It will cause a single file upload error to not stop the process, what I suggested would stop uploading the moment there was a problem.

mramato commented:
Just that one question about the catch block, other than that this is good to merge.  Thanks @ggetz

ggetz commented:
OK, in each catch block, I push the error to an array, and at the end I print the errors and fail the task. Currently, I'm just calling`done(1)`. Should I pass one of the errors?

mramato commented:
`done(1)` is fine.  Thanks!

mramato commented:
As soon as CI passes, I'll merge.

ggetz commented:
Thanks @mramato !

mramato commented on gulpfile.js:
```diff
@@ -353,8 +352,12 @@ function deployCesium(bucketName, uploadDirectory, cacheControl) {
+    var errors = [];
 
     return listAll(s3, bucketName, uploadDirectory, existingBlobs)
+    .catch(function(error) {
```
If `listAll` fails, I'm pretty sure we would just want to avoid the entire operation (which was the old behavior) but this catch makes that no longer the case.

ggetz replied:
Before this catch block was inside the `listAll` function itself, but I moved it out. Should I omit the catch block entirely?

mramato commented on gulpfile.js:
```diff
@@ -465,7 +468,21 @@ function deployCesium(bucketName, uploadDirectory, cacheControl) {
         .then(function() {
             console.log('Cleaned ' + existingBlobs.length + ' files.');
+        })
+        .catch(function(error) {
```
There's no reason for the catch here because it will bubble up to the parent promise.

mramato commented on gulpfile.js:
```diff
@@ -465,7 +468,21 @@ function deployCesium(bucketName, uploadDirectory, cacheControl) {
+            console.log(e);
+        });
+        done(1);
     });
```
You still need a catch at the top level of the promise.

mramato commented on gulpfile.js:
```diff
@@ -465,7 +468,21 @@ function deployCesium(bucketName, uploadDirectory, cacheControl) {
+        errors.map(function(e) {
+            console.log(e);
+        });
+        done(1);
```
I've bikeshedded enough so there's no need to change this, but I'm pretty sure you could have just returned `Promise.reject()` here instead of going pack to using the done parameter.

ggetz replied:
Would that work to fail the task since I'm not actually returning the promise at the task level?

mramato replied:
You would just return the promise at the task level, but like I said, no need to make any changes right now.

ggetz replied:
OK

mramato commented:
I actually had a few more minor comments, last ones, I swear :)

ggetz commented:
I fixed the catch blocks

mramato commented:
Thanks @ggetz, your travis deploy PR should now get pretty awesome.
